Fix: three not found view

kosmydel opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR adresses three things:
- fixes multiple not found views,
- fixes the back button,
- refactors the `WorkspaceMembersPage` to use the `WorkspacePageWithSections` component.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/34234
$ https://github.com/Expensify/App/issues/36266
PROPOSAL: N/A


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
- [ ] Verify that navigation works correctly.
- [ ] Verify the deleting workspace flow:
  - [ ] [Main device] Create new Workspace
  - [ ] [Main device] Navigate to Members
  - [ ] [Secondary device] Delete Workspace created in Main device
  - [ ] [Main device] Click on "Go back to home page" in LHN
- [ ] On Web (multiple not found views):
  - [ ] Go to the All Settings
  - [ ] Select a workspace
  - [ ] Change the workspace ID in the URL to a nonexistent one (e.g. `https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/profile`)
  - [ ] Verify that we have only one not-found view.
  - [ ] Verify that we can go back.
  - [ ] Verify other links:
    - [ ] Verify other pages:
      - [ ] https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/invoices
      - [ ] https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/reimburse
    - [ ] With RHP modal opened: 
      - [ ] https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/profile/name
      - [ ] https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/rateandunit
      - [ ] https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/rateandunit/unit
      - [ ] https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/invite
      - [ ] https://staging.new.expensify.com/settings/workspaces/kjashdjkashdaks/invite-message
      - [ ] Existing, without permissions: https://staging.new.expensify.com/settings/workspaces/E2BFF0D3DDCB1520/invoices

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
N/A

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
- [ ] Do the above testing steps

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/104823336/3a2167ef-e546-4f40-bed3-07cf328b7f69



</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->


https://github.com/Expensify/App/assets/104823336/1ddf643a-dad0-4084-9d89-6676501cd0af




</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->



https://github.com/Expensify/App/assets/104823336/fecb1ad3-dbf7-40bb-befb-2245d6e3ae83



</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->



https://github.com/Expensify/App/assets/104823336/6db5efc7-09c2-4f9d-8074-b6b29c8553fd



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/Expensify/App/assets/104823336/1437bc85-b664-451b-962f-0ae751e41a32


https://github.com/Expensify/App/assets/104823336/b122941a-6419-4207-ac5e-473ef92afa32




</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
<img width="1204" alt="Screenshot 2024-02-19 at 10 45 58" src="https://github.com/Expensify/App/assets/104823336/cabef320-932c-4e46-a084-3c119be155fe">

</details>

kosmydel committed:
fix three pane not found view

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel committed:
cleanup

kosmydel committed:
fix not authorized screen

kosmydel committed:
fix not found views - new approach

kosmydel committed:
cleanup

kosmydel committed:
fix going back from settings list

shawnborton commented:
Looks good to me, cc @trjExpensify @Expensify/design

kosmydel commented:
> Nice yeah, much better. On the first scenario where the RHP is open in the `MacOS: Chrome / Safari` video we do catch a glimpse of the modal sliding closed with the 404 being displayed. It's pretty fast though, so I'm not sure we need to worry about it too much.

@trjExpensify  I've noticed it too. During the navigation initialisation we don't know yet if the policy exists, so by the default, we mount the RHP. I just remove it after the Onyx initialises, when we got the full information. That's why we have this 'sliding' RHP.

trjExpensify commented:
>@trjExpensify I've noticed it too. During the navigation initialisation we don't know yet if the policy exists, so by the default, we mount the RHP. I just remove it after the Onyx initialises, when we got the full information. That's why we have this 'sliding' RHP.

Yeah, got it.. and that sounds like something that would be hard to resolve, so I think I'm cool with it. 👍

situchan commented:
I can take over.

There's conflict

trjExpensify commented:
@situchan assigned you. Please go ahead!

situchan commented:
Thanks. @kosmydel please conflict when you get a chance

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel committed:
cleanup

kosmydel commented:
@situchan  

Conflicts resolved! Sorry for the delay, I was OOO

situchan commented:
Conflicts again.

There are still issues on the pages with RHP modal


https://github.com/Expensify/App/assets/108292595/05d97fb7-72fd-46b3-8302-653fd459beaa

situchan commented on src/pages/workspace/WorkspacesListPage.tsx:
```diff
@@ -363,7 +363,7 @@ function WorkspacesListPage({policies, allPolicyMembers, reimbursementAccount, r
                     title={translate('common.workspaces')}
                     shouldShowBackButton={isSmallScreenWidth}
-                    onBackButtonPress={() => Navigation.goBack()}
+                    onBackButtonPress={() => Navigation.goBack(ROUTES.ALL_SETTINGS)}
```
What does this fix? This causes regression of wrong navigation animation

Repro step:
1. Visit https://dev.new.expensify.com:8082/settings/workspaces directly on mWeb
2. Click back button

main:

https://github.com/Expensify/App/assets/108292595/679c3c84-ac40-4638-a8fc-04db1519b37b


this branch:

https://github.com/Expensify/App/assets/108292595/a6ddf225-fc3a-4fa7-901a-256dc8a97e41

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

situchan commented:
> Hey @trjExpensify , @situchan
> 
> Do you think we have any case, where we want to display in the RHP not found/not authorized page and **do not** display it in the central pane?
> 
> If we don't, I could create much cleaner solution, and it would work probably in every case.

That might happen. i.e. RHP is page of editing any field and user is not allowed to edit it.

kosmydel committed:
fix: multiple not authorized pages, internal deeplinking

kosmydel commented:
Thanks @situchan!

I've resolved conflicts and fixed the reported issued. However, I still have to run the tests on all platforms (probably tomorrow).

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel commented:
Ok, I've tested it and it looks that it works.
cc @situchan

situchan commented:
@kosmydel please fix conflict

situchan commented on src/pages/workspace/WorkspaceInitialPage.tsx:
```diff
@@ -197,6 +197,20 @@ function WorkspaceInitialPage({policyDraft, policy: policyProp, policyMembers, r
+            return;
+        }
+        // We are dismissing any modals that are open when the NotFound view is shown
+        setTimeout(() => {
+            Navigation.dismissModal();
+        }, 0);
```
We usually avoid this kind of workaround. Isn't there any other way?

kosmydel replied:
Right!
I've changed it to the `isNavigationReady`. I still have to test it.

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel committed:
use isNavigationReady

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel commented:
Okay, a new issue was created to standardize workspace wrappers: https://github.com/Expensify/App/issues/37898

Now in this issue, we have two options:
1. Continue with this PR, test it, and merge it. The other issue will resolve issues on the remaining screens.
2. HOLD on the new issue.

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel committed:
support ideal nav v2

kosmydel commented:
I've adjusted it to the new ideal-nav v2. I've also added it to two new workspace wrappers to unblock this issue.

cc @situchan it is ready for review

kosmydel commented:
I've adjusted it to the new ideal-nav v2. I've also added it to two new workspace wrappers to unblock this issue.

cc @situchan it is ready for review

situchan commented:
> 
> However, I found out that two new screens in workspace settings got merged without using the `WorkspacePageWithSections` wrapper. I [asked the authors](https://github.com/Expensify/App/pull/37621/files#r1515780864) to confirm it was intentional, as my solution is partly based on this wrapper.

Was this also resolved?

kosmydel commented:
> > However, I found out that two new screens in workspace settings got merged without using the `WorkspacePageWithSections` wrapper. I [asked the authors](https://github.com/Expensify/App/pull/37621/files#r1515780864) to confirm it was intentional, as my solution is partly based on this wrapper.
> 
> Was this also resolved?

I've added this solution to those wrappers as well. The only thing I had to change here was adding the "shouldForceFullScreen" prop. So it's working fine on new pages.

situchan commented:
Bug: Not found page appears again after dismiss. Also happens on main so out of scope

tested with old link:
https://dev.new.expensify.com:8082/workspace/kjashdjkashdaks/profile


https://github.com/Expensify/App/assets/108292595/6fb6bb67-4393-476b-8e83-66f1654cb587

situchan commented:
@kosmydel please check this. Not blocker as not severe bug but can we fix this? Or out of scope?


https://github.com/Expensify/App/assets/108292595/a8c7bd99-5a10-42b4-bf98-418d0e3d1b85

situchan commented:
Another one on invite message page (similar to above but a bit different)


https://github.com/Expensify/App/assets/108292595/02beed7b-decb-42c6-83a9-1732afd7336b

situchan commented:
@kosmydel bump ^
please check above concerns. And please pull main.

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel commented:
> @kosmydel bump ^ please check above concerns. And please pull main.

Hey, sorry, I've been focusing on the Taxes feature. I'm going to work on this today.

kosmydel commented:
@situchan can you check it again? After pulling the main I can't reproduce those issues.

situchan commented:
Perf tests failing. Is this related?

situchan commented:
> @situchan can you check it again? After pulling the main I can't reproduce those issues.

@kosmydel still reproducible to me. Did you refresh?


https://github.com/Expensify/App/assets/108292595/fbf3fee7-e05c-4578-9533-fe520658ddcd

kosmydel commented:
Oh, I thought that the refresh was an issue. 

If we are talking about the quick glimpse of the not-found view before showing the loading indicator I think it is out of scope. The `withPolicyAndFullscreenLoading` displays it after one rerender.

![Screenshot 2024-03-21 at 12 09 39](https://github.com/Expensify/App/assets/104823336/46244420-5711-44a5-9616-ca70c25ec31e)

situchan commented:
Perf tests still failing

kosmydel committed:
Merge branch 'Expensify:main' into @kosmydel/duplicated-not-found-view

kosmydel commented:
> Perf tests still failing

Looks like merging main solved it

situchan commented:
@kosmydel this step is failing. It navigates to archived workspace chat instead of workspaces settings

- [x] [Main device] Click on "Go back to home page" in LHN
- [ ] Verify that the `Workspaces settings` is displayed.


https://github.com/Expensify/App/assets/108292595/fec5900a-69a4-42e5-b574-2a09505a68d6

situchan commented:
Maybe out of scope but there's inconsistency in back button behavior.

https://dev.new.expensify.com:8082/settings/workspaces/kjashdjkashdaks/invite goes to original chat
 https://dev.new.expensify.com:8082/settings/workspaces/kjashdjkashdaks/members goes to workspaces list page

I don't see any reason why they should be inconsistent.


https://github.com/Expensify/App/assets/108292595/efcaaaf6-66a9-4af7-b746-f253faf05d87

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel commented:
> @kosmydel this step is failing. It navigates to archived workspace chat instead of workspaces settings
> 
> * [x]  [Main device] Click on "Go back to home page" in LHN
> * [ ]  Verify that the `Workspaces settings` is displayed.

I believe that current behavior is acceptable. It has changed probably due to the ideal nav v2 changes.

kosmydel commented:
> Maybe out of scope but there's inconsistency in back button behavior.

This one is closely related to the navigation. When we open the RHP (`/invite`) the central pane is mounted simultaneously in the background. That's why we different navigation stack. I'm not sure if this one is worth further investigation.

situchan commented:
@kosmydel please merge main to fix TS error

kosmydel committed:
Merge branch 'Expensify:main' into @kosmydel/duplicated-not-found-view

Gonals commented:
@shawnborton, anything to add here?

shawnborton commented:
Where can I get the most up to date screenshots?

One thing that seemed weird to me was this particular screen from above:
![CleanShot 2024-03-29 at 17 45 40@2x](https://github.com/Expensify/App/assets/2319350/b75f7413-c35c-46ef-9305-1683fa414553)

blazejkustra commented:
@kosmydel is OOO today due to Easter, he should be able to answer and resolve conflicts tomorrow 😄

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel committed:
restart typecheck

kosmydel commented:
Hey, I've resolved the conflicts. 

> Where can I get the most up to date screenshots?
> 
> One thing that seemed weird to me was this particular screen from above
> I would expect that not found view to just take up the whole screen and not just the LHN.

@shawnborton I wasn't able to reproduce that one. I'm assuming, that during the recording of the video, the central pane data was loaded a bit later than the LHN. I'm not sure about the reason, might be something Onyx-related. 

Can you @situchan retest it to check if it still occurs for you?

shawnborton commented:
Friendly bump @situchan

kosmydel commented:
Just a quick update from me:

I've found out that @hayata-suenaga [is working on a higher-level approach](https://github.com/Expensify/App/issues/37898#issuecomment-2025563876) to the visibility of screens in the workspaces. This might resolve this issue as well, and with those changes, we could use a more holistic approach than this one.

hayata-suenaga commented:
@kosmydel, we decided to delay implementing the proposed refactoring as we faced some issues. We don't know when or even if we're going to proceed with this refactoring. Meantime, it might be a good idea to go ahead with merging this PR to fix the issue 😄

trjExpensify commented:
@situchan is out for a month, so @Gonals if you're happy with the latest changes. Maybe we can merge this one?

kosmydel committed:
Merge branch 'main' into @kosmydel/duplicated-not-found-view

kosmydel commented:
I've resolved conflicts and quickly tested. Before merging we should have another round of testing from C+ as the new architecture got merged. 

What should we do with this PR as @situchan is OOO?

Gonals commented:
Tis tests well for me, so I think we can go ahead and merge

kbecciv commented:
@kosmydel Could you kindly update the QA steps to reflect the staging URL?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8732363773) to staging by https://github.com/Gonals in version: 1.4.63-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8786104763) to production by https://github.com/mountiny in version: 1.4.63-21 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
